modules/packages/LinearAlgebra.chpl

```diff
@@ -0,0 +1,733 @@
+/* matrix-vector multiplication */
+private proc _matvecMult(A: [?Adom] ?eltType, X: [?Xdom] eltType)
+  where isBLASType(eltType)
+{
+  if Adom.rank != 2 || Xdom.rank != 1 then
+    compilerError("Rank sizes are not 2 and 1");
+
+  var Y: [Xdom] eltType;
+  gemv(A, X, Y, 1:eltType, 0:eltType);
+  return Y;
+}
+
+
+pragma "no doc"
+/* matrix-matrix multiplication */
+private proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
+  where isBLASType(eltType)
+{
+  if Adom.rank != 2 || Bdom.rank != 2 then
+    compilerError("Rank sizes are not 2");
+
+  var C: [Adom.dim(1), Bdom.dim(2)] eltType;
+  gemm(A, B, C, 1:eltType, 0:eltType);
+  return C;
+}
+
+
+/* Inner product of 2 vectors */
+proc inner(A: [?Adom], B: [?Bdom]) {
+  if Adom.rank != 1 || Bdom.rank != 1 then
+    compilerError("Rank sizes are not 1");
+
+  return + reduce(A[..]*B[..]);
+}
+
+
+/* Outer product of 2 vectors */
+proc outer(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  if Adom.rank != 1 || Bdom.rank != 1 then
+    compilerError("Rank sizes are not 1");
+
+  var C: [{Adom.dim(1), Bdom.dim(1)}] eltType;
+  forall (i,j) in C.domain do
+    C[i, j] = A[i]*B[j];
+  return C;
+}
+
+
+pragma "no doc"
+/* Generic matrix-vector multiplication */
+proc _matvecMult(A: [?Adom] ?eltType, X: [?Xdom] eltType)
+  where !isBLASType(eltType)
+{
+  if Adom.rank != 2 || Xdom.rank != 1 then
+    compilerError("Rank sizes are not 2 and 1");
+
+  var C: [Xdom] eltType;
+
+  // naive algorithm
+  forall i in Xdom do
+    C[i] = + reduce (A[i,..]*X[..]);
+
+  return C;
+}
+
+pragma "no doc"
+/* Generic matrix-matrix multiplication */
+proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
+  where !isBLASType(eltType)
+{
+  if Adom.rank != 2 || Bdom.rank != 2 then
+    compilerError("Rank sizes are not 2 and 1");
+
+  var C: [Adom.dim(1), Bdom.dim(2)] eltType;
+
+  // naive algorithm
+  forall (i,j) in C.domain do
+    C[i,j] = + reduce (A[i,..]*B[..,j]);
+
+  return C;
+}
+
+
+/* Return the matrix ``A`` to the ``bth`` power, where ``b`` is a positive
+   integral type. */
+proc matPow(A: [], b) where isNumeric(b) {
+  if !isIntegral(b) then
+    // TODO -- support all reals with Sylvester's formula
+    compilerError("matPow only support powers of integers");
+  if !isSquare(A) then
+    halt("Array not square");
+
+  return _expBySquaring(A, b);
+}
+
+
+pragma "no doc"
+/* Exponentiate by squaring recursively */
+private proc _expBySquaring(x: ?t, n): t {
+    // TODO -- _expBySquaring(pinv(x), -n);
+    if n < 0  then halt("Negative powers not yet supported");
+    else if n == 0  then return eye(x.domain);
+    else if n == 1  then return x;
+    else if n%2 == 0  then return _expBySquaring(dot(x, x), n / 2);
+    else return dot(x, _expBySquaring(dot(x, x), (n - 1) / 2));
+}
+
+
+/* Return element-wise power: ``a**B`` */
+proc matPow(a, B: [?Bdom] ?eltType) where isNumeric(a) {
+  var C: [Bdom] eltType = a**B;
+  return C;
+}
+
+
+/* Return element-wise exponential: ``e**A[i,j]`` */
```

Comment:
I've dropped the `matExp` routine for now, until it is implemented correctly as [matrix exponential](https://en.wikipedia.org/wiki/Matrix_exponential#Computing_the_matrix_exponential)